src/webgpu/shader/execution/expression/call/builtin/derivatives.ts

```diff
@@ -0,0 +1,218 @@
+    cases = packed.cases;
+    type = packed.resultType;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  // The two input values for a given case are distributed to two different invocations in a quad.
+  // We will populate a storage buffer with these input values laid out sequentially:
+  // [ case_0_input_1, case_0_input_0, case_1_input_1, case_1_input_0, ...]
+  //
+  // The render pipeline will be launched several times over a viewport size of (2, 2). Each draw
+  // call will execute a single quad (four fragment invocation), which will exercise two test cases.
+  // Each of these draw calls will use a different instance index, which is forwarded to the
+  // fragment shader. Each invocation will determine its index into the storage buffer using its
+  // fragment position and the instance index for that draw call.
+  //
+  // Consider two draw calls that test 4 cases (c_0, c_1, c_2, c_3).
+  //
+  // For derivatives along the 'x' direction, the mapping from fragment position to case input is:
+  // Quad 0: | c_0_i_1 | c_0_i_0 |     Quad 1: | c_2_i_1 | c_2_i_0 |
+  //         | c_1_i_1 | c_1_i_0 |             | c_3_i_1 | c_3_i_0 |
+  //
+  // For derivatives along the 'y' direction, the mapping from fragment position to case input is:
+  // Quad 0: | c_0_i_1 | c_1_i_1 |     Quad 1: | c_2_i_1 | c_3_i_1 |
+  //         | c_0_i_0 | c_1_i_0 |             | c_2_i_0 | c_3_i_0 |
+  //
+  ////////////////////////////////////////////////////////////////
+
+  // Determine the direction of the derivative ('x' or 'y') from the builtin name.
+  const dir = builtin[3];
+
+  // Determine the WGSL type to use in the shader, and the stride in bytes between values.
+  let valueStride = 4;
+  let wgslType = 'f32';
+  if (vectorize) {
+    wgslType = `vec${vectorize}f`;
+    valueStride = vectorize * 4;
+    if (vectorize === 3) {
+      valueStride = 16;
+    }
+  }
+
+  // Define a vertex shader that draws a quad over the full viewport, and a fragment shader that
+  // calls the derivative builtin with a value loaded from that fragment's index into the storage
+  // buffer (determined using the quad index and fragment position, as described above).
+  const code = `
+struct CaseInfo {
+  @builtin(position) position: vec4f,
+  @location(0) @interpolate(flat) quad_idx: u32,
+}
+
+@vertex
+fn vert(@builtin(vertex_index) vertex_idx: u32,
+        @builtin(instance_index) instance_idx: u32) -> CaseInfo {
+  const kVertices = array(
+    vec2f(-1, -1),
+    vec2f( 1, -1),
+    vec2f(-1,  1),
+    vec2f(-1,  1),
+    vec2f( 1, -1),
+    vec2f( 1,  1),
+  );
+  return CaseInfo(vec4(kVertices[vertex_idx], 0, 1), instance_idx);
+}
+
+@group(0) @binding(0) var<storage, read> inputs : array<${wgslType}>;
+@group(0) @binding(1) var<storage, read_write> outputs : array<${wgslType}>;
+
+@fragment
+fn frag(info : CaseInfo) {
+  let case_idx = u32(info.position.${dir === 'x' ? 'y' : 'x'} - 0.5);
```

Comment:
`u32(f32)` will round towards zero. By subtracting `-0.5` first, aren't we increasing the risk of quantizing to the wrong integer or low-precision devices?

Reply:
Good point, the subtraction is unnecessary and a potential hazard. Removed.